proof.ml

```diff
@@ -209,9 +208,50 @@ let from_sequent_and_rule_request_and_premises sequent rule_request premises =
     let expected_premises_conclusion = List.map get_conclusion (get_premises proof) in
     let given_premises_conclusion = List.map get_conclusion premises in
     if expected_premises_conclusion <> given_premises_conclusion then
-    raise (Technical_exception ("Premises conclusion do not match expected premises conclusion after applying rule"))
+    raise (Rule_exception (false, "Premises conclusion do not match expected premises conclusion after applying rule"))
     else set_premises proof premises;;
 
+(* AUTO REVERSE MODE *)
+
+exception NotApplicable;;
+
+let rec get_formula_position condition = function
+    | [] -> raise NotApplicable
+    | f :: tail -> if condition f then 0 else 1 + (get_formula_position condition tail);;
+
+let try_rule_request sequent rule_request =
+    try from_sequent_and_rule_request sequent rule_request
+    with Rule_exception _ -> raise NotApplicable;;
+
+let apply_reversible_rule proof =
+    let sequent = get_conclusion proof in
+    try try_rule_request sequent (Top (get_formula_position is_top sequent))
+        with NotApplicable ->
+    try try_rule_request sequent (Bottom (get_formula_position is_bottom sequent))
+        with NotApplicable ->
+    try try_rule_request sequent (Par (get_formula_position is_par sequent))
+        with NotApplicable ->
+    try try_rule_request sequent (Dereliction (get_formula_position is_double_whynot sequent))
+        with NotApplicable ->
+    try try_rule_request sequent (With (get_formula_position is_with sequent))
+        with NotApplicable ->
+    try try_rule_request sequent (Promotion (get_formula_position is_ofcourse sequent))
+        with NotApplicable ->
+    try try_rule_request sequent One
+        with NotApplicable ->
+    try if List.length sequent = 1 then try_rule_request sequent (Tensor 0) else raise NotApplicable
+        with NotApplicable ->
+    try try_rule_request sequent Axiom
+        with NotApplicable ->
+    proof;;
```

Comment:
@olaure01 I rewrote in the same order the auto-reversible rules this way.

Reply:
Looks good, I am doing tests.
I would like to add one more case: `?A` with context with `Zero`, `Dual` and `Litt` only, maps to `A, ?A` with same context (i.e. apply contraction and dereliction).
I was doing it with the old version, this one is much more readable, how simple is it to do it here?